File: NetbeansProjects/AStar/src/astar/AStar.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package astar;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import priorityqueue.PriorityQueue;
import priorityqueue.PriorityQueueImpl;

/**
 *
 * @author Tobias NOTE: Need to create a customary priority queue, since
 * remove(Object o) in java is based on equals and therefore is O(n).
 */
public class AStar
{

    public static <N extends Node<N>> Iterable<N> getShortestPath(N start, N goal, Heuristic<N> heuristic)
    {
        Map<N, NodeData> nodeDataMap = new HashMap<>();
        PriorityQueue<NodeData> openSet = new PriorityQueueImpl<>(new Order());
        Set<N> closedSet = new HashSet<>();

        NodeData<N> curNode = new NodeData<>(start);
        curNode.g = 0;
        nodeDataMap.put(start, curNode);
        while (true)
        {
            for (Edge<N> edge : curNode.node)
            {
                N other = edge.getEnd();
                if (!closedSet.contains(other))
                {
                    boolean inOpenSet = true;
                    NodeData<N> otherNode = nodeDataMap.get(other);
                    if (otherNode == null)
                    {
                        otherNode = new NodeData<>(other);
                        nodeDataMap.put(other, otherNode);
                        inOpenSet = false;
                    }
                    float newG = edge.getCost() + curNode.g;
                    if (otherNode.g > newG)
                    {
                        otherNode.g = newG;
                        if (inOpenSet)
                        {
                            openSet.update(otherNode);
                        }
                        otherNode.prev = curNode;
                    }
                    if (!inOpenSet)
                    {
                        otherNode.h = heuristic.getMinimumDist(other, goal);
                        openSet.add(otherNode);
                    }
                }
            }
            if (openSet.isEmpty())
            {
                return null;
            }
            closedSet.add(curNode.node);
            curNode = openSet.poll();
            if (curNode.node.equals(goal))
            {
                ArrayList<N> res = new ArrayList<>();
                do
                {
                    res.add(curNode.node);
                    curNode = curNode.prev;
                } while (curNode != null);
                Collections.reverse(res);
                return res;
            }
        }
    }

    private static class NodeData<N>
    {

        public N node;
        public float g;
        public float h;
        public NodeData prev;

        public NodeData(N node)
        {
            this.node = node;
            this.g = Float.MAX_VALUE;
            this.h = 0;
            this.prev = null;
        }

        @Override
        public int hashCode()
        {
            return node.hashCode();
        }

        @Override
        public boolean equals(Object obj)
        {
            if (this == obj)
            {
                return true;
            }
            if (obj == null)
            {
                return false;
            }
            if (getClass() != obj.getClass())
            {
                return false;
            }
            final NodeData<N> other = (NodeData<N>) obj;
            return Objects.equals(this.node, other.node);
        }
    }

    private static class Order implements Comparator<NodeData>
    {

        @Override
        public int compare(NodeData d1, NodeData d2)
        {
            float f1 = d1.g + d1.h;
            float f2 = d2.g + d2.h;
            if (f1 < f2)
            {
                return -1;
            }
            if (f1 > f2)
            {
                return 1;
            }
            if(d1.h < d2.h)
            {
                return -1;
            }
            if(d1.h > d2.h)
            {
                return 1;
            }
            return 0;
        }
    }

}

File: VisualStudioProjects/GraphAlgorithms/GraphAlgorithms/Impl/AStar.cs
using System;
using System.Collections.Generic;

namespace GraphAlgorithms
{
    class AStar
    {
        public static IEnumerable<IPathNode<N>> GetShortestPath<N>(N start, N goal, IHeuristic<N> heuristic) where N : INode<N>
        {
            C5.IDictionary<N, PathNode<N>> nodeDictionary = new C5.HashDictionary<N, PathNode<N>>();
            C5.IPriorityQueue<PathNode<N>> openSet = new C5.IntervalHeap<PathNode<N>>(new PathNodeComparer<N>(), C5.MemoryType.Normal);
            C5.ICollection<N> closedSet = new C5.HashSet<N>();


            PathNode<N> curNode = new PathNode<N>(start);
            curNode.g = 0;
            nodeDictionary.Add(start, curNode);
            while (true)
            {
                foreach(IEdge<N> edge in curNode.node)
                {
                    N other = edge.GetEnd();
                    if (!closedSet.Contains(other))
                    {
                        PathNode<N> otherNode = null;
                        if (!nodeDictionary.Find(ref other, out otherNode))
                        {
                            otherNode = new PathNode<N>(other);
                            nodeDictionary.Add(other, otherNode);
                        }
                        float newG = edge.GetCost() + curNode.g;
                        if (otherNode.g > newG)
                        {
                            otherNode.g = newG;
                            if (otherNode.queueHandle != null)
                            {
                                openSet.Replace(otherNode.queueHandle, otherNode);
                            }
                            otherNode.prev = curNode;
                        }
                        if (otherNode.queueHandle == null)
                        {
                            otherNode.h = heuristic.MinDist(other, goal);
                            C5.IPriorityQueueHandle<PathNode<N>> handle = null;
                            openSet.Add(ref handle, otherNode);
                            otherNode.queueHandle = handle;
                        }
                    }
                }
                if (openSet.IsEmpty)
                {
                    return null;
                }
                closedSet.Add(curNode.node);
                curNode = openSet.DeleteMin();
                if (curNode.node.Equals(goal))
                {
                    C5.ArrayList<IPathNode<N>> res = new C5.ArrayList<IPathNode<N>>(C5.MemoryType.Normal);
                    do
                    {
                        res.Add(curNode);
                        curNode = curNode.prev;
                    } while (curNode != null);
                    res.Reverse();
                    return res;
                }
            }
        }


        private class PathNode<N> : IPathNode<N>, IEquatable<PathNode<N>> where N : INode<N>
        {
            public N node { get; }
            public PathNode<N> prev { get; set; }
            public C5.IPriorityQueueHandle<PathNode<N>> queueHandle { get; set; }
            public float g { get; set; }
            public float h { get; set; }

            public PathNode(N node)
            {
                this.node = node;
                this.prev = null;
                this.queueHandle = null;
                this.g = float.MaxValue;
                this.h = 0;
            }

            public float GetCost()
            {
                return g;
            }

            public N GetNode()
            {
                return node;
            }

            public IPathNode<N> GetPrevious()
            {
                return prev;
            }

            public bool Equals(PathNode<N> other)
            {
                // If parameter is null, return false.
                if (Object.ReferenceEquals(other, null))
                {
                    return false;
                }
                return this.node.Equals(other.node);
            }

            public override int GetHashCode()
            {
                return node.GetHashCode();
            }
        }

        private class PathNodeComparer<N> : IComparer<PathNode<N>> where N : INode<N>
        {
            public int Compare(PathNode<N> x, PathNode<N> y)
            {
                float fx = x.g + x.h;
                float fy = y.g + y.h;
                if (fx < fy) return -1;
                if (fx > fy) return 1;
                if (x.h < y.h) return -1;
                if (x.h > y.h) return 1;
                return 0;
            }
        }
    }
}
